src/libraries/Microsoft.Extensions.FileSystemGlobbing/src/MatcherExtensions.cs

```diff
@@ -51,9 +51,18 @@ public static void AddIncludePatterns(this Matcher matcher, params IEnumerable<s
         /// <returns>Absolute file paths of all files matched. Empty enumerable if no files matched given patterns.</returns>
         public static IEnumerable<string> GetResultsInFullPath(this Matcher matcher, string directoryPath)
         {
-            IEnumerable<FilePatternMatch> matches = matcher.Execute(new DirectoryInfoWrapper(new DirectoryInfo(directoryPath))).Files;
-            string[] result = matches.Select(match => Path.GetFullPath(Path.Combine(directoryPath, match.Path))).ToArray();
+            PatternMatchingResult patternMatchingResult = matcher.Execute(new DirectoryInfoWrapper(new DirectoryInfo(directoryPath)));
+            if (!patternMatchingResult.HasMatches)
+            {
+                return Array.Empty<string>();
+            }
 
+            IEnumerable<FilePatternMatch> matches = patternMatchingResult.Files;
+            List<string> result = matches is ICollection matchCollection ? new(matchCollection.Count) : new();
```

Comment:
> matches is ICollection

`ICollection<T>` doesn't inherit `ICollection`; many of our collection types implement both, but it's not guaranteed. Is this just a "good enough" thing because we expect most inputs would be e.g. `T[]` or `List<T>`?

Reply:
Judging by the [source](https://source.dot.net/#Microsoft.Extensions.FileSystemGlobbing/Internal/MatcherContext.cs,cea2df48faa28307,references), `matches` will always be a `List<T>` here.

Reply:
@eerhardt is my understanding correct? If so, to reduce complexity a bit, can we remove the `new()` fallback and maybe add an "is ICollection" assertion?

Reply:
Will merge the change without this since its still valuable.

Reply:
Sorry I missed the original question.

Since this is a public API and `Matcher.Execute` is `virtual`, the `PatternMatchingResult` can have any `IEnumerable<FilePatternMatch> Files` set on it. It isn't guaranteed to always be a `List<T>`.

Reply:
> Is this just a "good enough" thing because we expect most inputs would be e.g. T[] or List<T>?

Yes, our implementation always returns a `List<T>`, so we typically expect this to be `List<T>` in most cases.